### PR TITLE
Add a new ``time_support`` context manager for plotting times

### PR DESCRIPTION
**Pull request by astrofrog**
Matplotlib natively provides a mechanism for plotting dates and times on one or both of the axes, as described in [Date tick labels](https://matplotlib.org/3.1.0/gallery/text_labels_and_annotations/date.html). To make use of this, users need to explicitly access ``plot_date`` on time instances they want to plot. While this might be ok for a number of situations, I thought it would be nice to have a way of more seamlessly using times when plotting that gives more control over the time scale and format.

This PR implements a ``time_support`` context manager/function that is meant to be similar to ``quantity_support``:

```python
from matplotlib import pyplot as plt
from astropy.time import Time
from astropy.visualization import time_support

time_support()

plt.figure(figsize=(5,3))
plt.plot(Time([58000, 59000, 62000], format='mjd'), [1.2, 3.3, 2.3])
```

matplotlib_integration-3

(by default the time is removed from the ISO string if it's always midnight for all tick labels, this can be controlled with the ``simplify`` argument).

This can be used as a context manager and can take the format and scale to use for the axis:

```python
with time_support(format='mjd', scale='tai'):
    plt.figure(figsize=(5,3))
    plt.plot(Time([58000, 59000, 62000], format='mjd'))
```

matplotlib_integration-5

Note that this also allows ``Time`` to be used in ``set_xlim/set_ylim`` for example. One of the benefits of this approach is that all times are automatically converted to a common scale/format.

A few notes:

* Something which I'm not entirely happy with, which also affects ``quantity_support``, is that even when using the context manager, once an axis has been used inside the context manager, it has a cache of the converter and it's impossible to change it again later. I think this is something that would need to be fixed in MPL though. As long as a new axis/figure is created, then the context manager works fine.

* I'm also not really fond of the current names ``quantity_support`` and ``time_support``, I feel there should be a ``plot`` or ``mpl`` in there somewhere so it's clear that it's related to plotting.

* I've consolidated the docs to mention ``quantity_support`` and ``time_support`` in one place, in the visualization docs.

* There is still some cleanup to do in various places and I think I can expand the docs and tests a little, but I don't want to invest too much time until I hear whether people are interested in this. I need this anyway for another project so I decided to implement it here, but it's ok if people would prefer not to include it!

### CODE REVIEW

File: astropy/visualization/time.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+from astropy.visualization.wcsaxes.utils import select_step_hour, select_step_scalar
+
+from matplotlib.ticker import MaxNLocator, ScalarFormatter
+
+__all__ = ['time_support']
+
+YMDHMS_FORMATS = ('fits', 'iso', 'isot')
+
+
+def time_support(scale='utc', format='isot', simplify=True):
```
Comment (mhvk) [inline astropy/visualization/time.py]: Would it be possible to have the default set by the times first used? (say, `scale=None` here)

File: astropy/visualization/time.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        if self._remove:
+            del units.registry[Time]
+
+    def convert(self, value, unit, axis):
+        """
+        Convert a Time value to a scalar or array.
+        """
+        scaled = getattr(value, self.scale)
+        if self.format in YMDHMS_FORMATS:
+            return scaled.mjd
```
Comment (mhvk) [inline astropy/visualization/time.py]: Is this done to make the values numerical? But, if so, shouldn't it happen for all string formats? For best precision for relevant times, one might consider using `cxcsec` or so.

File: docs/visualization/matplotlib_integration.rst [hunk 3]
```diff
@@ -0,0 +1,121 @@
+.. _quantity:
```
Comment (mhvk) [inline docs/visualization/matplotlib_integration.rst]: This link seems wrong.

File: astropy/visualization/time.py (hunk 1 above)
Comment (astrofrog) [reply astropy/visualization/time.py]: Agree, this will be consistent with ``quantity_support`` - fixed

File: astropy/visualization/time.py (hunk 2 above)
Comment (astrofrog) [reply astropy/visualization/time.py]: I've now fixed compatibility for all string formats (and have tests for all of them).

File: docs/visualization/matplotlib_integration.rst (hunk 3 above)
Comment (astrofrog) [reply docs/visualization/matplotlib_integration.rst]: Removed

**Comment by astrofrog** [comment]
@mhvk - thanks for the initial review! You mentioned that using MJD might not be ideal, so I was wondering whether, since you are intimately familiar with astropy.time, you would mind trying out this PR and finding cases where the precision doesn't work properly?

One use case I know of already that needs to be fixed is when looking at very small time intervals, the string-based formats need to have their precision updated (they are currently fixed to millisecond precision). But it would be great if you could let me know if you can find other cases that won't work properly/optimally!

**Comment by astrofrog** [comment]
This whole functionality doesn't work very well with Matplotlib 2.0 just because of the way that it dealt with non-numpy arrays back then (it split the Time array into a list of many times, and same for Quantity). Since Matplotlib 2.1, 2.2, 3.0 and 3.1 are out now and 2.0 is over two years old, maybe we can just drop it? (rather than have to add convoluted workarounds here).

**Comment by larrybradley** [comment]
The `travis` tests failures look real.

**Comment by mhvk** [comment]
@astrofrog - using `mjd` means the best possible precision is
```
np.spacing(Time.now().mjd * u.day).to(u.us)
# <Quantity 0.62864274 us>
```
This is probably good enough for most purposes...  

But it brings up a different question: do you also handle `TimeDelta`? I think those should just be treated as `Quantity`...

**Comment by astrofrog** [comment]
@mhvk - just to check, do you mean we should support it in ``quantity_support`` or in ``time_support``? It would be much cleaner to support it in the former. I guess this wouldn't matter if we then introduce a context manager such as ``matplotlib_astropy_support`` which enables all options like you suggest.

**Comment by mhvk** [comment]
@astrofrog - yes, you're right, `TimeDelta` is more logically supported in `quantity_support`.

I do think an overall `astropy_support` makes sense; one really shouldn't want one without the other. Indeed, in implementing support for the new `__array_function__` (which allows *all* numpy functions to work on quantities!), I'm finding that plain `plot(q1, q2)` starts breaking (because quantities are no longer interpreted as plain ndarray). So, the recommendation may well become to always have `astropy_support` on when plotting.

**Comment by astrofrog** [comment]
@mhvk @larrybradley - this is now ready for a final review (I think ``TimeDelta`` and the joint context manager should be handled in a separate PR)

File: astropy/visualization/time.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+            if self.scale is None:
+                self.scale = x.scale
+            return 'astropy_time'
+
+        def convert(self, value, unit, axis):
+            """
+            Convert a Time value to a scalar or array.
+            """
+
+            # For Matplotlib < 2.2
```
Comment (mhvk) [inline astropy/visualization/time.py]: Should we increase our minimum version?

Comment (astrofrog) [reply astropy/visualization/time.py]: I already just increased it to 2.1 from 2.0, but we could consider increasing further especially given the 2 year support for 4.0. I wish we had a written down policy for this :)

Comment (mhvk) [reply astropy/visualization/time.py]: Probably OK to leave for later: https://github.com/astropy/astropy/issues/8822#issuecomment-502203917

File: astropy/visualization/time.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+      >>> from astropy import units as u
+      >>> from astropy import visualization
+      >>> with visualization.time_support():  # doctest: +IGNORE_OUTPUT
+      ...     plt.figure()
+      ...     plt.plot(Time(['2016-03-22T12:30:31', '2016-03-22T12:30:38', '2016-03-22T12:34:40']))
+      ...     plt.draw()
+
+    Parameters
+    ----------
+    format : str, optional
```
Comment (larrybradley) [inline astropy/visualization/time.py]: `scale` should come before `format` to match the function signature.

Comment (astrofrog) [reply astropy/visualization/time.py]: Fixed, and I also made the function signature have keyword-only arguments.

**Comment by astrofrog** [comment]
@mhvk issues opened in https://github.com/astropy/astropy/issues/8859 and https://github.com/astropy/astropy/issues/8860

**Comment by mhvk** [comment]
Thanks for raising the issues. SInce @larrybradley's comment was addressed, and he was happy otherwise, I'll merge now - very nice to have this!!

**Comment by mhvk** [comment]
Hmm, having done it, not sure that the right way is to "dismiss a review"....

**Pull request merged**